lib/charms/mongodb/v0/set_status.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+"""Code for handing statuses in the app and unit."""
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from ops.charm import CharmBase
+from ops.framework import Object
+from ops.model import StatusBase
+
+from config import Config
+
+# The unique Charmhub library identifier, never change it
+LIBID = "9b0b9fac53244229aed5ffc5e62141eb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class MongoDBStatusHandler(Object):
+    """Verifies versions across multiple integrated applications."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+    ) -> None:
+        """Constructor for CrossAppVersionChecker.
+
+        Args:
+            charm: charm to inherit from
+            version: (int), the current version of the desired attribute of the charm
+            relations_to_check: (List), a list of relations who should have compatible versions
+                with the current charm
+            version_validity_range: (Optional Dict), a list of ranges for valid version ranges.
+                If not provided it is assumed that relations on the provided interface must have
+                the same version.
+        """
+        super().__init__(charm, None)
+        self.charm = charm
+
+        # TODO Future PR: handle update_status
+
+    # BEGIN Helpers
+
+    def set_and_share_status(self, status: StatusBase):
+        """Sets the charm status and shares to app status and config-server if applicable."""
+        # TODO Future Feature/Epic: process other statuses, i.e. only set provided status if its
+        # appropriate.
+        self.charm.unit.status = status
+
+        self.set_app_status()
+
+        if self.charm.is_role(Config.Role.SHARD):
+            self.share_status_to_config_server()
+
+    def set_app_status(self):
+        """TODO Future Feature/Epic: parse statuses and set a status for the entire app."""
+
+    def are_charm_units_all_active_or_waiting_for_upgrade(self) -> bool:
+        """Returns True if all charm units status's show that they are ready for upgrade."""
+        goal_state = self.charm.model._backend._run(
+            "goal-state", return_output=True, use_json=True
+        )
+        for _, unit_state in goal_state["units"].items():
+            if unit_state["status"] == "active":
+                continue
+            if unit_state["status"] != "waiting":
+                return False
+
+            if not self.charm.get_cluster_mismatched_revision_status():
```

Comment:
This only runs for the current unit

Reply:
Yes, but it should have the same result across all units so it should be no big deal :)

Reply:
But since the output is the same, it makes little sense to keep it inside the loop :)

Reply:
Do you mean take `self.charm.get_cluster_mismatched_revision_status()` out of the loop?

I could be wrong but I think the value of `self.charm.get_cluster_mismatched_revision_status()` will be cached and it shouldnt hurt the performance